mule-extensions-api/src/main/java/org/mule/runtime/extension/internal/loader/enricher/ExtensionTypesDeclarationEnricher.java

```diff
@@ -135,6 +138,13 @@ private void registerType(ExtensionDeclarer declarer, MetadataType type) {
 
       @Override
       public void visitObject(ObjectType objectType) {
+        objectType.getAnnotation(TypeIdAnnotation.class).ifPresent(typeId -> {
+          if (typeId.getValue().equals(messageTypeId)) {
+            for (ObjectFieldType field : objectType.getFields()) {
+              field.accept(this);
+            }
+          }
+        });
         declarer.withType(objectType);
```

Comment:
Shouldn't we stop doing this when the object type is a message type?

Reply:
I left it as it was to preserve old behavior as well. It doesn't really change anything because the MessageType is not registrable and does not have an OpenRestriction.

Reply:
Neither. If you know that this is MessageMetadataType, then you should cast it, obtain the Payload and Attribute metadata types and run them through the same visitor

Reply:
Would't that imply going back to the previous fix which involved doing "isinstanceof" and casting? The point of using the annotation is to avoid casting and prevent potential issues with serialization.

Reply:
No. Don't loose sight of what you're actually trying to accomplish here. The functionality to implement is to add the payload and attributes types to the extension's dictionary.

The original review was not only about the `instanceof`, it was also mainly about the fact that what you were not doing that correctly, independently of the casting and instanceof. This comment is about that still being done incorrectly. 

As for casting and instanceof, ideal solution is to have that resolved by the visitor, but since that's not possible you don't have a choice but to do that. What you have to make sure is to then do thiings correctly once that situations is detected